Address incosistent behaviour in dump_smt function across solvers

mihaistate opened the pull request:
Updated the z3_convt::output_smt() and the bitwulza_convt::dump_smt() functions to print
to stdout when no output was defined

Improved the handling of empty and "-" output paths in the smtlib_conv::dump_smt() function

Addressed issue #2167

mihaistate05 committed:
Improved SMT solvers

Updated the z3_convt::output_smt() and the bitwulza_convt::dump_smt() functions to print
to stdout when no output was defined

Signed-off-by: mihaistate05 <mihaita.state@yahoo.com>

mihaistate committed:
Improved smtlib_conv::dump_smt() function

Improved the handling of empty and "-" output paths

Signed-off-by: mihaistate <mihai.state03@yahoo.com>

mihaistate committed:
Improved smtlib_conv::dump_smt() function

Improved the handling of empty and "-" output paths

Signed-off-by: mihaistate05 <mihai.state03@yahoo.com>

mihaistate committed:
Update .gitignore to exclude build artifacts

rafaelsamenezes commented on src/.gitignore:
```diff
@@ -22,3 +22,19 @@ stamp-h1
+CMakeFiles/
+CPack*
+cmake_install.cmake
+Makefile
```
Could you either revert this or remove the duplicates?

mihaistate replied:
I will remove the duplicates.

rafaelsamenezes commented on src/solvers/z3/z3_conv.h:
```diff
@@ -210,8 +210,9 @@ class z3_convt : public smt_convt,
-
+  
 public:
+  virtual void output_smt();
```
Is thi method defined?

mihaistate replied:
Yes, the method is defined here: https://github.com/esbmc/esbmc/commit/5dbc705a45d99793cefbe0c0daa6e5cc3ed50785#r159957012

rafaelsamenezes reviewed (changes requested):
Overall its ok. Just asking for some cleanup.

XLiZHI commented:
https://github.com/esbmc/esbmc/blob/ec6e9d78b67a63c8d6b683a59a2332ea6b76bfe2/src/solvers/cvc5/cvc5_conv.h#L156

I think the function return type of CVC5 still needs to be changed.

mihaistate committed:
Solved some SMT related errors

Changed the return type of the dump_smt() function
for consistency reasons
Changed the dump_smt() function type in the bitwuzla_conv
file and modified it to satisfy both compiler and logging needs

Signed-off-by: mihaistate05 <mihai.state03@yahoo.com>

mihaistate commented:
@XLiZHI I changed the function.

lucasccordeiro commented:
@mihaistate: Your PR is failing in our CI. Could you please fix the issues?

mihaistate committed:
Updated the return type of dump_smt() function for bitwulza and boolector solvers

Changed the return type to string for the dump_smt() function
inside the bitwulza_conv.cpp and boolector_conv.cpp files
for having consistency across all solvers
Improved these functions to return appropiate messages and
work well with files

Signed-off-by: mihaistate05 <mihai.state03@yahoo.com>

mihaistate commented:
@lucasccordeiro I think I fixed the CI errors with my last commit.

lucasccordeiro commented:
@mihaistate: Our CI is still failing with your last commit. Could you please check the issues?

lucasccordeiro commented:
@mihaistate: https://app.codacy.com/gh/esbmc/esbmc/pull-requests/2524/issues

mihaistate committed:
Changed the cvc5_convt::dump_smt() function to string

The cvc5_convt::dump_smt() function now returns the SMT
output
Changed a parameter to a match a file declaration in the
bitwuzla_convt::dump_smt() function, at the
bitwuzla_print_formula(bitw, "smt2", file, 2) function call

Signed-off-by: mihaistate05 <mihai.state03@yahoo.com>

mihaistate commented:
I have an error which tells me that Github Action is trying to reference my feature branch from my fork into the main branch: fatal: invalid reference: feature/improve-smt-output. This is because the git-auto-commit action is trying to push changes to the feature branch, but Github Action does not have permission to push to a fork.  To fix this error, I would need someone to edit the workflow file (https://github.com/esbmc/esbmc/actions/runs/15654290748/workflow?pr=2524) at lines 94-95 to skip the auto-commit step: 

- uses: stefanzweifel/git-auto-commit-action@v5
  if: ${{ github.event.pull_request.head.repo.full_name == github.repository }}
  with:
    commit_message: "Auto-format code

This is the full error message: fatal: invalid reference: feature/improve-smt-output
Error: Invalid status code: 128
    at ChildProcess.<anonymous> (/home/runner/work/_actions/stefanzweifel/git-auto-commit-action/v5/index.js:17:19)
    at ChildProcess.emit (node:events:524:28)
    at maybeClose (node:internal/child_process:1104:16)
    at ChildProcess._handle.onexit (node:internal/child_process:304:5) {
  code: 128
}
Error: Invalid status code: 128
    at ChildProcess.<anonymous> (/home/runner/work/_actions/stefanzweifel/git-auto-commit-action/v5/index.js:17:19)
    at ChildProcess.emit (node:events:524:28)
    at maybeClose (node:internal/child_process:1104:16)
    at ChildProcess._handle.onexit (node:internal/child_process:304:5)

rafaelsamenezes commented:
> I have an error which tells me that Github Action is trying to reference my feature branch from my fork into the main branch: fatal: invalid reference: feature/improve-smt-output. This is because the git-auto-commit action is trying to push changes to the feature branch, but Github Action does not have permission to push to a fork.  To fix this error, I would need someone to edit the workflow file (https://github.com/esbmc/esbmc/actions/runs/15654290748/workflow?pr=2524) at lines 94-95 to skip the auto-commit step: 
> 
> 
> 
> - uses: stefanzweifel/git-auto-commit-action@v5
> 
>   if: ${{ github.event.pull_request.head.repo.full_name == github.repository }}
> 
>   with:
> 
>     commit_message: "Auto-format code
> 
> 
> 
> This is the full error message: fatal: invalid reference: feature/improve-smt-output
> 
> Error: Invalid status code: 128
> 
>     at ChildProcess.<anonymous> (/home/runner/work/_actions/stefanzweifel/git-auto-commit-action/v5/index.js:17:19)
> 
>     at ChildProcess.emit (node:events:524:28)
> 
>     at maybeClose (node:internal/child_process:1104:16)
> 
>     at ChildProcess._handle.onexit (node:internal/child_process:304:5) {
> 
>   code: 128
> 
> }
> 
> Error: Invalid status code: 128
> 
>     at ChildProcess.<anonymous> (/home/runner/work/_actions/stefanzweifel/git-auto-commit-action/v5/index.js:17:19)
> 
>     at ChildProcess.emit (node:events:524:28)
> 
>     at maybeClose (node:internal/child_process:1104:16)
> 
>     at ChildProcess._handle.onexit (node:internal/child_process:304:5)

I think you can enable edits by reviewers. That should be enough to allow the commit. If not, you can just execute clang format locally on your machine (just be sure of using the correct version).

lucasccordeiro commented:
@mihaistate: I have just invited you to join the ESBMC github repo: https://github.com/esbmc/esbmc/invitations

mihaistate committed:
Solved Codacy code analysis checks

lucasccordeiro commented:
@mihaistate: codacy is complaining about this include: cstdio.

mihaistate commented:
Thanks, @lucasccordeiro. This library is needed for the FILE declaration (fopen and fclose methods) inside this function for bitwuzla_conv.cpp:
std::string bitwuzla_convt::dump_smt()
{
  const std::string &path = options.get_option("output");

  // Print formulas using binary bit-vector output format
  if (path.empty() || path == "-") {
  bitwuzla_print_formula(bitw, "smt2", stdout, 2);
  }
  else 
  {
    FILE* file = fopen(path.c_str(), "w");
    if (!file)
    {
      log_error("Could not open output file '{}'", path);
      return "Failed to open output file.";
    }
    bitwuzla_print_formula(bitw, "smt2", file, 2);
    fclose(file);
  }
  log_status("SMT formula dumped successfully");
  return "SMT formula dumped successfully";

}   A solution would be that a maintainer of the project could "Bypass status checks" for Codacity on this PR.  Alternatively, I could do this at the beginning of the bitwuzla_conv.cpp file: 
**// cppcheck-suppress missingIncludeSystem
#include cstdio**    
I think this latter one could be also relevant, but maybe is not the most elegant one.

XLiZHI commented:
> A solution would be that a maintainer of the project could "Bypass status checks" for Codacity on this PR

That's okay, I think we can ignore that.

As for the C++ code style check, you can fix it by running the following code locally on `ESBMC_Project/esbmc` (you may need to change the clang path) :

`find src -iname *.h -o -iname *.c -o -iname *.cpp -o -iname *.hpp -iname *.hh | grep -v src/ansi-c/cpp | grep -v src/clang-c-frontend/headers/ | grep -v src/c2goto/library/libm/musl | xargs .././clang11/bin/clang-format -style=file -i -fallback-style=none`

mihaistate committed:
Changed the clang format

mihaistate commented:
I ran the command: find src -iname '*.h' -o -iname '*.c' -o -iname '*.cpp' -o -iname '*.hpp' -o -iname '*.hh'   | grep -v src/ansi-c/cpp   | grep -v src/clang-c-frontend/headers/   | grep -v src/c2goto/library/libm/musl   | xargs clang-format-14 -style=file -i -fallback-style=none in ESBMC/esbmc and pushed the changes in my last commit, but the error still persists.

XLiZHI commented:
> I ran the command: find src -iname '_.h' -o -iname '_.c' -o -iname '_.cpp' -o -iname '_.hpp' -o -iname '*.hh' | grep -v src/ansi-c/cpp | grep -v src/clang-c-frontend/headers/ | grep -v src/c2goto/library/libm/musl | xargs clang-format-14 -style=file -i -fallback-style=none in ESBMC/esbmc and pushed the changes in my last commit, but the error still persists.

Try clang11

Or you can create your own branch in the esbmc repository, so when you commit the PR, the git action will automatically change the format. I believe you have the authority to do this.

mihaistate commented:
@XLiZHI I fetched the initial feature branch from my fork to a new one on the esbmc project. Now, all checks pass, I would need a review from @rafaelsamenezes and also the initial review that request changes from last week that needs to pass, to enable the merge request. I am also updating the branch frequently with the master one.

lucasccordeiro commented:
Thanks for submitting this PR, @mihaistate.

@XLiZHI and @rafaelsamenezes : Thank you for your support here!

Anthonysdu commented:
@mihaistate Mathsat solver needs to be addressed as well, our benchexec CI is failing: https://github.com/esbmc/esbmc/actions/runs/15862503043/job/44722745878
